Support connections to Unix domain sockets

apeschar opened the pull request:
It's not currently possible to connect to a Redis server that uses Unix domain sockets.

Using Unix domain sockets is handy because it allows using filesystem permissions to control access, and because there's no need to worry about conflicting ports.

This PR introduces support for Unix domain sockets by setting the socket path in the `host` and setting `port` to `0`.

I've taken this approach instead of extending the `connection_spec` or turning it into a variant type because I suspect a lot of code uses `connection_spec` directly and this would be incompatible.

(I think it would be ideal if we supported a single string (`redis://` URL) to specify connection information, as this makes managing client configuration a lot easier and allows extending supported connection types and options without having to change the code that uses this library. But I think that's for another day.)

c-cube commented on .gitignore:
```diff
@@ -8,5 +8,5 @@
 *.exe
-
 /result*
+/socket/redis.sock
```
```suggestion
*.sock
```

c-cube commented on src/s.ml:
```diff
@@ -8,7 +8,8 @@ module type IO = sig
 
-  val connect : string -> int -> fd t
+  val getaddrinfo : string -> string -> Unix.getaddrinfo_option list -> Unix.addr_info list t
+  val connect : Unix.socket_domain -> Unix.sockaddr -> fd t
```
maybe we keep the old `connect` and add a `connect_full` or `connect_socket` function?

apeschar replied:
My assumption was that this signature is only used internally, and therefore there's no need to maintain backward compatibility.

Since I've now moved the logic which calls getaddrinfo out of the IO modules to the client, I'm not sure how to move it back without duplicating the logic.

As I see it, it makes sense to keep the API surface of the IO modules as small as possible to reduce the amount of work when editing or adding new IO modules.

c-cube commented on src/s.ml:
```diff
@@ -86,7 +87,8 @@ module type Client = sig
+      Specify a path and port 0 to use unix domain sockets.
+      @param port Port to connect to (default [6379])
       @since 0.5 *)
 
```
```suggestion
val connection_spec_unix_socket : string -> connection_spec
```

more explicit :)

c-cube reviewed (approved):
Looks good, only a few details to iron out :)

apeschar commented:
I rebased and added the connection_spec_unix_socket.

Have a look at my comment on your suggestion to see if that makes sense or if I'm still missing something. :)

c-cube commented:
thank you! :)

apeschar commented:
Thank you for merging! 🎉

c-cube commented:
No worries 😁. It's also released on opam!
